File: src/lib/geocode.ts
/**
 * Geocoding using geocoder.imtools.info — a self-hosted service backed
 * by HDX/OCHA COD boundary data.
 *
 * Reverse geocode: GET /geocode?lat=&lon= returns ADM0–ADM4 P-codes and names.
 * Forward geocode: GET /geocode?address= returns lat/lon + ADM0–ADM4 P-codes.
 *
 * Output keys use the `_geo_adm{n}_pcode` / `_geo_adm{n}_name` convention that
 * the rest of the pipeline (hook.ts → editSubmission) expects.
 */

const GEOCODER_URL = "https://geocoder.imtools.info/geocode";

type GeocoderResponse = {
  success: boolean;
  error?: string;
  [key: string]: unknown;
};

// ── Public API ────────────────────────────────────────────────────────────────

/**
 * Returns a map of `_geo_adm{n}_pcode` / `_geo_adm{n}_name` fields ready to be
 * written back to the Kobo submission via `editSubmission()`.
 *
 * Returns {} for uncovered coordinates or on error (logged, not thrown).
 */
export async function geocodeSubmission(lat: number, lon: number): Promise<Record<string, string>> {
  const params = new URLSearchParams({ lat: String(lat), lon: String(lon) });
  let data: GeocoderResponse;
  try {
    const res = await fetch(`${GEOCODER_URL}?${params}`, {
      headers: { "User-Agent": "kobo2logie/1.0" },
    });
    if (res.status === 404) {
      console.log(`[geo] No coverage at (${lat}, ${lon})`);
      return {};
    }
    if (!res.ok) {
      console.error(`[geo] HTTP ${res.status} from geocoder`);
      return {};
    }
    data = await res.json<GeocoderResponse>();
  } catch (err) {
    console.error(`[geo] fetch error: ${err}`);
    return {};
  }

  if (!data.success) {
    console.log(`[geo] Geocoder returned success=false: ${data.error}`);
    return {};
  }

  // Map adm{n}_pcode / adm{n}_name → _geo_adm{n}_pcode / _geo_adm{n}_name
  const out: Record<string, string> = {};
  for (let n = 0; n <= 4; n++) {
    const pcode = data[`adm${n}_pcode`];
    const name  = data[`adm${n}_name`];
    if (typeof pcode === "string" && pcode) out[`_geo_adm${n}_pcode`] = pcode;
    if (typeof name  === "string" && name)  out[`_geo_adm${n}_name`]  = name;
  }
  return out;
}

/**
 * Forward geocode an address string to lat/lon + P-codes.
 *
 * Returns a map containing `_geo_latitude`, `_geo_longitude`, and the same
 * `_geo_adm{n}_pcode` / `_geo_adm{n}_name` fields as `geocodeSubmission()`.
 *
 * Returns {} when the address cannot be resolved or on error (logged, not thrown).
 */
export async function geocodeAddress(address: string): Promise<Record<string, string>> {
  const params = new URLSearchParams({ address });
  let data: GeocoderResponse;
  try {
    const res = await fetch(`${GEOCODER_URL}?${params}`, {
      headers: { "User-Agent": "kobo2logie/1.0" },
    });
    if (res.status === 404) {
      console.log(`[geo/address] No result for address: ${address}`);
      return {};
    }
    if (!res.ok) {
      console.error(`[geo/address] HTTP ${res.status} from geocoder`);
      return {};
    }
    data = await res.json<GeocoderResponse>();
  } catch (err) {
    console.error(`[geo/address] fetch error: ${err}`);
    return {};
  }

  if (!data.success) {
    console.log(`[geo/address] success=false: ${data.error}`);
    return {};
  }

  const out: Record<string, string> = {};

  // Include resolved coordinates
  if (typeof data.latitude === "number") out["_geo_latitude"] = String(data.latitude);
  if (typeof data.longitude === "number") out["_geo_longitude"] = String(data.longitude);

  for (let n = 0; n <= 4; n++) {
    const pcode = data[`adm${n}_pcode`];
    const name  = data[`adm${n}_name`];
    if (typeof pcode === "string" && pcode) out[`_geo_adm${n}_pcode`] = pcode;
    if (typeof name  === "string" && name)  out[`_geo_adm${n}_name`]  = name;
  }
  return out;
}

File: app/lib/geocode.py
"""
Geocoding via geocoder.imtools.info.
Port of src/lib/geocode.ts.
"""

import logging

import httpx

logger = logging.getLogger(__name__)

GEOCODER_URL = 'https://geocoder.imtools.info/geocode'
HEADERS = {'User-Agent': 'kobo2logie/2.0'}


def geocode_submission(lat: float, lon: float) -> dict[str, str]:
    """
    Reverse geocode coordinates → ADM0–ADM4 P-codes and names.
    Returns keys like _adm0_pcode, _adm0_name, etc.
    Returns {} on error or no coverage.
    """
    try:
        resp = httpx.get(
            GEOCODER_URL,
            params={'lat': str(lat), 'lon': str(lon)},
            headers=HEADERS,
            timeout=10,
        )
        if resp.status_code == 404:
            logger.info('[geo] No coverage at (%s, %s)', lat, lon)
            return {}
        if not resp.is_success:
            logger.error('[geo] HTTP %s from geocoder', resp.status_code)
            return {}
        data = resp.json()
    except Exception as exc:
        logger.error('[geo] fetch error: %s', exc)
        return {}

    if not data.get('success'):
        logger.info('[geo] success=false: %s', data.get('error'))
        return {}

    out: dict[str, str] = {}
    for n in range(5):
        pcode = data.get(f'adm{n}_pcode')
        name = data.get(f'adm{n}_name')
        if isinstance(pcode, str) and pcode:
            out[f'_adm{n}_pcode'] = pcode
        if isinstance(name, str) and name:
            out[f'_adm{n}_name'] = name
    return out


def geocode_address(address: str) -> dict[str, str]:
    """
    Forward geocode an address string → lat/lon + P-codes.
    Returns keys like _latitude, _longitude, _adm0_pcode, etc.
    Returns {} when the address cannot be resolved or on error.
    """
    try:
        resp = httpx.get(
            GEOCODER_URL,
            params={'address': address},
            headers=HEADERS,
            timeout=10,
        )
        if resp.status_code == 404:
            logger.info('[geo/address] No result for: %s', address)
            return {}
        if not resp.is_success:
            logger.error('[geo/address] HTTP %s from geocoder', resp.status_code)
            return {}
        data = resp.json()
    except Exception as exc:
        logger.error('[geo/address] fetch error: %s', exc)
        return {}

    if not data.get('success'):
        logger.info('[geo/address] success=false: %s', data.get('error'))
        return {}

    out: dict[str, str] = {}
    if isinstance(data.get('latitude'), (int, float)):
        out['_latitude'] = str(data['latitude'])
    if isinstance(data.get('longitude'), (int, float)):
        out['_longitude'] = str(data['longitude'])
    for n in range(5):
        pcode = data.get(f'adm{n}_pcode')
        name = data.get(f'adm{n}_name')
        if isinstance(pcode, str) and pcode:
            out[f'_adm{n}_pcode'] = pcode
        if isinstance(name, str) and name:
            out[f'_adm{n}_name'] = name
    return out
